### PR TITLE
feat: Attempt to dump a stacktrace to stderr prior to process abort on SIGSEGV/SIGILL/SIGBUS

### PR DESCRIPTION
**Pull request by alamb**
# Rationale:
In our internal system (see https://github.com/influxdata/conductor/issues/413) the `influxdb_iox` process is existing with `SIGSEGV` (aka segfaulting for some reason). 

While a core dump is likely the ideal solution in such a case, it is not always available and/or the logistics of getting such a dump trace is challenging (e.g. if the resident memory size is 10s of GB so would be the core file and not all systems are prepared to create / save such a file).

Thus,  having the process itself try and dump out as much information prior to its death is useful

Also known as "its getting real now"

# Changes:
1. As suggested by @mkmik  https://github.com/influxdata/conductor/issues/413#issuecomment-889409811,  add a signal handler for SIGSEGV and SIGILL that dumps a backtrace of the current thread and exits

FWIW other systems I have worked on all had something like this and it worked pretty well in practice 🤷 

# Notes:
I fully acknowledge this code is totally unsafe, and as memory is corrupted somehow there is no guarantee we'll get a stack dump. However, since the process is about to abort anyways it doesn't really matter if we happen to also segfault / blow up in the handler. In practice I have almost always been able to get a stack trace with a mechanism such as this

# Testing
I tested this manually like this:

```shell
cargo run -- run -v  --object-store=file --data-dir=$HOME/.influxdb_iox --server-id=42
```

And then sent the `influxdb_iox` process a SEGSEGV via

```shell
alamb@MacBook-Pro influxdb_iox2 % ps -ef | grep iox
ps -ef | grep iox
  501 73165 40013   0 Tue08AM ttys000    0:00.10 ./target/debug/influxdb_iox sql
  501 61518 69256   0 11:59AM ttys001    0:05.08 kubectl -n iox port-forward svc/iox-query-3000 1234:8082 1235:8080
  501 64664 61414   0  4:39PM ttys002    0:01.91 target/debug/influxdb_iox run -v --object-store=file --data-dir=/Users/alamb/.influxdb_iox --server-id=42
  501 64687 61441   0  4:39PM ttys005    0:00.00 grep iox

alamb@MacBook-Pro influxdb_iox2 % kill -SEGV 64664
kill -SEGV 64664
alamb@MacBook-Pro influxdb_iox2 %
```

Which results in:

```

cd /Users/alamb/Software/influxdb_iox2 && cargo run -- run -v  --object-store=file --data-dir=$HOME/.influxdb_iox --server-id=42
    Finished dev [unoptimized + debuginfo] target(s) in 0.37s
     Running `target/debug/influxdb_iox run -v --object-store=file --data-dir=/Users/alamb/.influxdb_iox --server-id=42`
Jul 29 16:39:07.814  INFO influxdb_iox::influxdb_ioxd: InfluxDB IOx server starting git_hash="67abb09a4975da86d99d386627f2d12f947c80b7" num_cpus=16 build_malloc_conf="\u{0}"
...
Jul 29 16:39:30.592 ERROR influxdb_iox::influxdb_ioxd::http: Error while handling request error=HandleRequest(DatabaseNotFound { name: "26f7e5a4b7be365b_917b97a92e883afc" }, "/api/v2/write/") error_message="A route was unable to handle the request for target: /api/v2/write/" method=POST uri=/api/v2/write?bucket=917b97a92e883afc&org=26f7e5a4b7be365b span_id=None content_length=None
Signal 11, Stack trace for thread "main"
   0: backtrace::backtrace::libunwind::trace
             at /Users/alamb/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.61/src/backtrace/libunwind.rs:90:5
      backtrace::backtrace::trace_unsynchronized
             at /Users/alamb/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.61/src/backtrace/mod.rs:66:5
   1: backtrace::backtrace::trace
             at /Users/alamb/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.61/src/backtrace/mod.rs:53:14
   2: backtrace::capture::Backtrace::create
             at /Users/alamb/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.61/src/capture.rs:176:9
   3: backtrace::capture::Backtrace::new
             at /Users/alamb/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.61/src/capture.rs:140:22
   4: influxdb_iox::signal_handler
             at src/main.rs:281:59
   5: __sigtramp


Compilation abort trap: 6 at Thu Jul 29 16:39:33
```

### CODE REVIEW

**Review by NGA-TRAN — APPROVED** [review]

**Comment by alamb** [comment]
@tustvold  points out we already have a singal handler here: https://github.com/influxdata/influxdb_iox/blob/alamb%2Fdump_stacktrace_as_a_last_goodbye/src/influxdb_ioxd.rs#L87

It is handling SIGINT and SIGTERM, so it *shouldn't* interfere with the handling of SIGSEGV / SIGILL but it seems potentially problematic.

**Comment by alamb** [comment]
🤔  when I tried to use the tokio system to handle `SIGSEGV` it got huffy on me:

```diff
diff --git a/src/influxdb_ioxd.rs b/src/influxdb_ioxd.rs
index 3b0ece05..cacf8990 100644
--- a/src/influxdb_ioxd.rs
+++ b/src/influxdb_ioxd.rs
@@ -84,11 +84,15 @@ pub type Result<T, E = Error> = std::result::Result<T, E>;
 /// This method returns if either are signalled
 #[cfg(unix)]
 async fn wait_for_signal() {
+    use sig::ffi::Sig;
+
     use tokio::signal::unix::{signal, SignalKind};
+    let mut segv = signal(SignalKind::from_raw(Sig::SEGV)).expect("failed to register signal handler");
     let mut term = signal(SignalKind::terminate()).expect("failed to register signal handler");
     let mut int = signal(SignalKind::interrupt()).expect("failed to register signal handler");
 
     tokio::select! {
+        _ = segv.recv() => info!("Received SIGSEGV"),
         _ = term.recv() => info!("Received SIGTERM"),
         _ = int.recv() => info!("Received SIGINT"),
     }
diff --git a/src/main.rs b/src/main.rs
index c901f9cf..7e94701c 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -137,7 +137,7 @@ enum Command {
 }
 
 fn main() -> Result<(), std::io::Error> {
-    install_crash_handler(); // attempt to render a useful stacktrace to stdout
+    //install_crash_handler(); // attempt to render a useful stacktrace to stdout
 
     // load all environment variables from .env before doing anything
     load_dotenv();
```

```

    Finished dev [unoptimized + debuginfo] target(s) in 53.58s
     Running `target/debug/influxdb_iox run -v --object-store=file --data-dir=/Users/alamb/.influxdb_iox --server-id=42`
Jul 29 17:28:33.152  INFO influxdb_iox::influxdb_ioxd: InfluxDB IOx server starting git_hash="71b96aced758ec07e72687591cbe68c055f682e4" num_cpus=16 build_malloc_conf="\u{0}"
Jul 29 17:28:33.153  INFO influxdb_iox::influxdb_ioxd: Using File for object storage
Jul 29 17:28:33.156  INFO influxdb_iox::influxdb_ioxd: gRPC server listening bind_address=127.0.0.1:8082
Jul 29 17:28:33.156  INFO influxdb_iox::influxdb_ioxd: HTTP server listening bind_address=127.0.0.1:8080
Jul 29 17:28:33.157  INFO influxdb_iox::influxdb_ioxd: InfluxDB IOx server ready git_hash="71b96aced758ec07e72687591cbe68c055f682e4"
Jul 29 17:28:33.157 ERROR panic_logging: thread 'main' panicked at 'UNKNOWN', src/influxdb_ioxd.rs:90:60
thread 'main' panicked at 'failed to register signal handler: Custom { kind: Other, error: "Refusing to register signal 11" }', src/influxdb_ioxd.rs:90:60
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

Compilation exited abnormally with code 101 at Thu Jul 29 17:28:33
```

**Review by tustvold — APPROVED** [review]
LGTM - I only mentioned the existing handler as a way to avoid additional dependencies. If the tokio handler doesn't let us use it then I think this is fine. Might be worth adding a comment, to avoid future confusion as to why we have two different locations registering signal handlers

**Comment by alamb** [comment]
Let me dig around to see if we really need the `sig` package -- we certainly don't need a new crate to get a few numbers (`11`) lol

File: src/main.rs [hunk 1]
```diff
@@ -256,3 +258,30 @@ fn load_dotenv() {
...
+}
+
+unsafe extern "C" fn signal_handler(sig: i32) {
+    use backtrace::Backtrace;
+    use std::process::abort;
+    let name = std::thread::current()
+        .name()
+        .map(|n| format!(" for thread \"{}\"", n))
+        .unwrap_or_else(|| "".to_owned());
+    println!(
```
Comment (mkmik) [inline src/main.rs:280]: Can we make it print the stack trace to stderr instead (eprintln)?

Comment (alamb) [reply src/main.rs]: In 62fa571a

**Comment by alamb** [comment]
FYI @tustvold  I rewrote this code directly in terms of `libc` (rather than `sig` which simply wrapped the libc calls)` in https://github.com/influxdata/influxdb_iox/pull/2155/commits/a90e56224d5245018922cf728dd591b51ed27c08

Now I think there are no additional dependencies

File: src/main.rs [hunk 2]
```diff
@@ -137,6 +137,8 @@ enum Command {
 }
 
 fn main() -> Result<(), std::io::Error> {
+    install_crash_handler(); // attempt to render a useful stacktrace to stdout
```
Comment (alamb) [inline src/main.rs:140]:
```suggestion
    install_crash_handler(); // attempt to render a useful stacktrace to stderr
```

**Review by mkmik — APPROVED** [review]

**Comment by mkmik** [comment]
awesome, thanks!

**Review by tustvold — APPROVED** [review]

**Pull request merged**